### PR TITLE
Incorporate TinyASM into Firedrake

### PR DESCRIPTION
**Pull request by JDBetteridge**
# Description
Move TinyASM into Firedrake. Later job is to rewrite in C, but for now keep as C++ and use pybind11.

`python -c "import tinyasm"` segfaults, but this happened before :slightly_frowning_face:

### CODE REVIEW

File: tinyasm/matinvert.cpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#include <petsc.h>
+#include <petscblaslapack.h>
+
+#include <numeric>
+#include <vector>
+using namespace std;
```
Comment (JHopeCollins) [inline tinyasm/matinvert.cpp:6]:
```suggestion
```

File: tinyasm/tinyasm.cpp [hunk 2]
```diff
@@ -0,0 +1,452 @@
+#include <iostream>
```
Comment (JHopeCollins) [inline tinyasm/tinyasm.cpp:1]:
```suggestion
```

File: tinyasm/tinyasm.cpp [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+#include <numeric>
+#include <chrono>
```
Comment (JHopeCollins) [inline tinyasm/tinyasm.cpp:3]:
```suggestion
```

File: tinyasm/tinyasm.cpp [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+                    for(int i=0; i<dof; i++){
+                        for(int j=0; j<dof; j++){
+                            cout << vvv[i*dof+j] << " ";
+                        }
+                        cout << endl;
+                    }
+                    ierr = MatDenseRestoreArrayRead(localmats[p],&vvv);CHKERRQ(ierr);
+                    cout << endl;
+                }
+            }
```
Comment (JHopeCollins) [inline tinyasm/tinyasm.cpp:124]:
```suggestion
```

**Review by JHopeCollins — CHANGES_REQUESTED** [review]

**Comment by wence-** [comment]
TODO (perhaps)? Ask @florianwechsung what license is appropriate. I made some non-substantive changes that I am happy to license under whatever.

**Comment by dham** [comment]
Good catch Lawrence. We need to ask Florian for a licence. LGPL would match the rest of Firedrake. MIT would also be fine.

**Comment by florianwechsung** [comment]
Happy to see TinyASM being moved into firedrake! LGPL is good with me.

**Review by JHopeCollins — APPROVED** [review]
Firedrake is under LGPL so no extra license needed.

**Review by dham — APPROVED** [review]

**Pull request merged**